In [61]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import cv2
from PIL import Image
import os
from skimage import color
from skimage.feature import hog
from sklearn import svm
from sklearn.neighbors import KNeighborsClassifier # using 1NN

# 1. Download the [cow teat datasets](https://github.com/YoushanZhang/SCTL) (10 points) resize image to (224, 224)

### (1). Create a train data loader that returns image arrays and labels
### (2). Create a test data loader that returns image arrays and file names
### (3). Print image arrays, labels and file names dimensions 

### (1). Create a train data loader that returns image arrays and labels


In [62]:
data_dir_train= "/Users/sayedraheel/Desktop/Computer Vision Assign/Train"

In [ ]:
data_dir_train= "/Users/sayedraheel/Desktop/Computer Vision Assign/Train"
data_dir_test="/Users/sayedraheel/Desktop/Computer Vision Assign/Test"

In [63]:
data_dir_test="/Users/sayedraheel/Desktop/Computer Vision Assign/Test"

### Create a train data loader that returns image arrays and labels

In [64]:
#Creating a train data loader that returns image arrays and labels
def load_data2(data_dir):
    images = []
    labels = []
    
    # Define a dictionary to map label names to their corresponding numeric values
    label_dict = {'Score_1': 0, 'Score_2': 1, 'Score_3': 2,'Score_4':3}
    
    # Iterate over the directories in the data directory
    for dir_name in os.listdir(data_dir):
        dir_path = os.path.join(data_dir, dir_name)
        
        # Skip any non-directory files
        if not os.path.isdir(dir_path):
            continue
        
        # Get the label for this directory based on its name
        label = label_dict.get(dir_name, None)
        if label is None:
            continue
        
        # Iterate over the files in the directory
        for file_name in os.listdir(dir_path):
            file_path = os.path.join(dir_path, file_name)
            
            # Skip any non-image files
            if not file_path.endswith('.jpg'):
                continue
            
            # Read in the image file and append it to the images list along with its label
            image = cv2.imread(file_path)
            image = cv2.resize(image, (224,224))
            images.append(image)
            labels.append(label)
            
    return print('Image array dimensions of Train Dataset:', np.shape(images)), print('Label dimensions:', np.shape(labels))


In [65]:
#calling the function
load_data2(data_dir_train)

Image array dimensions of Train Dataset: (1149, 224, 224, 3)
Label dimensions: (1149,)


(None, None)

In [66]:
#getting the labels of Train data
def getting_labels(data_dir):
    images = []
    labels = []
    
    # Define a dictionary to map label names to their corresponding numeric values
    label_dict = {'Score_1': 0, 'Score_2': 1, 'Score_3': 2,'Score_4':3}
    
    # Iterate over the directories in the data directory
    for dir_name in os.listdir(data_dir):
        dir_path = os.path.join(data_dir, dir_name)
        
        # Skip any non-directory files
        if not os.path.isdir(dir_path):
            continue
        
        # Get the label for this directory based on its name
        label = label_dict.get(dir_name, None)
        if label is None:
            continue
        
        # Iterate over the files in the directory
        for file_name in os.listdir(dir_path):
            file_path = os.path.join(dir_path, file_name)
            
            # Skip any non-image files
            if not file_path.endswith('.jpg'):
                continue
            
            # Read in the image file and append it to the images list along with its label
            image = cv2.imread(file_path)
            image = cv2.resize(image, (224,224))
            images.append(image)
            labels.append(label)
            
    return labels


In [67]:
#getting the labels
train_labels= getting_labels(data_dir_train)


### Create a test data loader that returns image arrays and file names

In [68]:
# Creating a test data loader that returns image arrays and file names

def test_data_loader(data_dir):
    images = []
    file_names = []
    
    # Iterate over the files in the directory
    for file_name in os.listdir(data_dir):
        file_path = os.path.join(data_dir, file_name)
        
        # Skip any non-image files
        if not file_path.endswith('.jpg'):
            continue
        
        # Read in the image file and append it to the images list
        image=cv2.imread(file_path)
        image = cv2.resize(image, (224,224))
        images.append(image)
        
        # Append the file name to the file names list
        file_names.append(file_name)
    
    return print('Image array dimensions of Test Dataset:', np.shape(images)), file_names[:5]


In [69]:
test_data_loader(data_dir_test)

Image array dimensions of Test Dataset: (380, 224, 224, 3)


(None,
 ['20200724_094839_LH_C1_PP_M.jpg',
  'R7_0665_17_crop_RH_C3_PN_m.jpg',
  '20200724_094909_LF_C1_PP_M.jpg',
  '20200724_095044_RF_C1_PN_M.jpg',
  'R9_0779_41_crop_RF_C2_PN_m.jpg'])

### Print image arrays, labels and file names dimensions

In [70]:
import os
import cv2

def Test_load_data(data_dir):
    images = []
    labels = []
    file_names = []

    # Iterate over the directories in the data directory
    for dir_name in os.listdir(data_dir):
        dir_path = os.path.join(data_dir, dir_name)

        # Skip any non-directory files
        if not os.path.isdir(dir_path):
            continue

        # Iterate over the files in the directory
        for file_name in os.listdir(dir_path):
            file_path = os.path.join(dir_path, file_name)
            file_names.append(file_name)

            # Skip any non-image files
            if not file_path.endswith('.jpg'):
                continue

            # Read in the image file and append it to the images list
            image=cv2.imread(file_path)
            image = cv2.resize(image, (224,224))
            images.append(image)

            # Append the label to the labels list
            labels.append(dir_name)

    # Combine the images, labels, and file names into a list of tuples
    data = list(zip(images, labels, file_names))

    # Print the dimensions of the images, labels, and file names
    print("DIMENSION:", np.shape(images))
    print("Dimension of combined images with their labels", np.shape(data))
    print("File_Name:", file_names[:1])

    # Return the data
    return data


In [71]:
#calling the function
Test_load_data(data_dir_train)

DIMENSION: (1149, 224, 224, 3)
Dimension of combined images with their labels (1149, 3)
File_Name: ['20200724_094453_RF_C1_PP_M.jpg']


/Users/sayedraheel/opt/anaconda3/envs/dlenv/lib/python3.6/site-packages/numpy/core/_asarray.py:83: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  return array(a, dtype, copy=False, order=order)


[(array([[[ 49,  66, 109],
          [ 52,  69, 112],
          [ 46,  63, 106],
          ...,
          [132, 137, 187],
          [133, 138, 187],
          [130, 136, 184]],
  
         [[ 49,  66, 109],
          [ 53,  70, 113],
          [ 50,  67, 110],
          ...,
          [127, 132, 181],
          [131, 136, 185],
          [131, 136, 184]],
  
         [[ 46,  63, 106],
          [ 50,  67, 110],
          [ 50,  67, 110],
          ...,
          [124, 128, 178],
          [131, 136, 185],
          [132, 138, 186]],
  
         ...,
  
         [[ 56,  68,  86],
          [ 55,  67,  85],
          [ 53,  65,  83],
          ...,
          [ 43,  53,  71],
          [ 44,  54,  72],
          [ 44,  54,  72]],
  
         [[ 58,  70,  88],
          [ 59,  71,  89],
          [ 59,  71,  89],
          ...,
          [ 42,  52,  70],
          [ 44,  54,  72],
          [ 45,  55,  73]],
  
         [[ 53,  65,  83],
          [ 57,  69,  87],
          [ 56,  68,  86

# 2. Extract features of training and test images using HOG (20 points)
Please print the size of extracted features, e.g., training features: 1149 * d, test features: 380 *d

### Feature extraction for training Data

In [72]:
# Extracting features of training  images using HOG
import os
import cv2
import numpy as np
from skimage.feature import hog

def load_data_hog(data_dir):
    images_for_hog = []
    labels = []
    file_names = []
    for dir_name in os.listdir(data_dir):
        dir_path = os.path.join(data_dir, dir_name)
        if not os.path.isdir(dir_path):
            continue
        for file_name in os.listdir(dir_path):
            file_path = os.path.join(dir_path, file_name)
            if not file_path.endswith('.jpg'):
                continue
            img = cv2.imread(file_path)
            img = cv2.resize(img, (512,512))
            images_for_hog.append(img)
            labels.append(dir_name)
            file_names.append(file_name)
    return images_for_hog

def extract_features(images):
    features = []
    for image in images:
        hog_features = hog(image, orientations=9, cells_per_block=(2, 2), multichannel=True)
        features.append(hog_features)
    return features



In [76]:
#saving train images in images_for_hog_train
Train_images=load_data_hog(data_dir_train)

In [77]:
hog_features_train=extract_features(Train_images)
hog_features_train_shape=np.shape(hog_features_train)
x=hog_features_train_shape[0]
y=hog_features_train_shape[1]

In [78]:
#calling function that extract features using HOG for Train images
#Extract_features=extract_features(Train_images)
print("Feature of Train images",x,"*",y)
#print("Total Feature of Train image:",Extract_features[1]*len(Train_images))

Feature of Train images 1149 * 142884


### Feature Extraction for Test data

In [79]:
#fucntion to get test images resize and save it in images_for
def test_image(data_dir):
    test_images_for = []
    
    # Iterate over the files in the directory
    for file_name in os.listdir(data_dir):
        file_path = os.path.join(data_dir, file_name)
        
        # Skip any non-image files
        if not file_path.endswith('.jpg'):
            continue
        
        # Read in the image file and append it to the images list
        img = cv2.imread(file_path)
        img = cv2.resize(img, (512,512))
        test_images_for.append(img)
            
    return test_images_for

def extract_features_test(images):
    features = []
    for image in images:
        hog_features = hog(image, orientations=9, cells_per_block=(2, 2), multichannel=True)
        features.append(hog_features)
    return features




In [80]:
#calling function that gets all the test image in one place and save it in "images_for_hog_test"
Test_images=test_image(data_dir_test)

In [81]:
hog_features_test=extract_features(Test_images)
hog_features_test_shape=np.shape(hog_features_test)
x=hog_features_test_shape[0]
y=hog_features_test_shape[1]

In [82]:
hog_features_test

[array([0.2406468 , 0.15149423, 0.08091403, ..., 0.10124609, 0.14724496,
        0.05106609]),
 array([0.05022964, 0.        , 0.0252141 , ..., 0.26493926, 0.16630751,
        0.09307344]),
 array([0.23429266, 0.10855431, 0.21914633, ..., 0.1434239 , 0.08704411,
        0.01733868]),
 array([0.29409854, 0.        , 0.02130137, ..., 0.18996569, 0.13207379,
        0.14273213]),
 array([0.23751412, 0.04159816, 0.19588219, ..., 0.10888126, 0.25385113,
        0.15648933]),
 array([0.23961908, 0.19492659, 0.08659569, ..., 0.03389069, 0.07144785,
        0.14751779]),
 array([0.27446576, 0.18985664, 0.11786304, ..., 0.07895911, 0.11033592,
        0.17881544]),
 array([0.28225796, 0.28225796, 0.12352448, ..., 0.07876823, 0.25681836,
        0.14038114]),
 array([0.24937319, 0.17387509, 0.12371429, ..., 0.2528149 , 0.11424551,
        0.        ]),
 array([0.30290456, 0.30290456, 0.30290456, ..., 0.21333878, 0.0864369 ,
        0.05300445]),
 array([0.28880617, 0.01826992, 0.        , ..., 0

In [83]:
#calling function that extract features using HOG for test images 
#Extract_feature=extract_features(Test_images)
print("Feature of Test images",x,"*",y)
#print("Total Feature of Train image:",Extract_feature[1]*len(Test_images))

Feature of Test images 380 * 142884


In [84]:
#filepath for test and train dataset
data_dir_train= "/Users/sayedraheel/Desktop/Computer Vision Assign/Train"
data_dir_test="/Users/sayedraheel/Desktop/Computer Vision Assign/Test"

# 3. Extract features of training and test images using SIFT (20 points)
Please print the size of extracted features, e.g., training features: 1149 * d, test features: 380 *d

In [85]:
#images extracted and loaded in the jupyter as "Train_images", "Test_images"
print("Test_images Size",len(Test_images))
print("Train_images Size",len(Train_images))

Test_images Size 380
Train_images Size 1149


In [86]:

import cv2

# Function to extract SIFT features from training images
def extract_sift_train(train_images):
    # Create SIFT object
    sift = cv2.xfeatures2d.SIFT_create()
    # Create an empty list to hold the descriptors of each image
    descriptors_list = []

    # Loop over each image in the training set
    for image in train_images:
        # Detect keypoints and compute SIFT descriptors for the image
        kp, des = sift.detectAndCompute(image, None)
        # If the descriptors are not empty, add them to the list
        if des is not None:
            descriptors_list.append(des)

    # Concatenate the descriptors of all images in the list into a single numpy array
    descriptors_train = np.concatenate(descriptors_list, axis=0)

    # Print the size of the training features and the total number of features
    print("Training features:",len(train_images),'*', np.shape(descriptors_train))

    # Return the training descriptors
    return descriptors_train

# Function to extract SIFT features from test images
def extract_sift_test(test_images):
    # Create SIFT object
    sift = cv2.xfeatures2d.SIFT_create()
    # Create an empty list to hold the descriptors of each image
    descriptors_list = []

    # Loop over each image in the test set
    for image in test_images:
        # Detect keypoints and compute SIFT descriptors for the image
        kp, des = sift.detectAndCompute(image, None)
        # If the descriptors are not empty, add them to the list
        if des is not None:
            descriptors_list.append(des)

    # Concatenate the descriptors of all images in the list into a single numpy array
    descriptors_test = np.concatenate(descriptors_list, axis=0)
    
    # Print the size of the test features and the total number of features
    print("Test features:",len(test_images),'*',np.shape(descriptors_test))


    # Return the test descriptors
    return descriptors_test


In [87]:
#the number of detected keypoints in  input images,
#and y part of tuple is  the length of each descriptor vector for each keypoint. 
#and each keypoint has a descriptor vector of length x.

In [88]:
#calling the fucntion to extract SIFT features in Training Dataset
sift_features_train=extract_sift_train(Train_images)
sift_features_train_shape=np.shape(sift_features_train)

Training features: 1149 * (154971, 128)


In [89]:
#calling the fucntion to extract SIFT features in Test Dataset
sift_features_test=extract_sift_test(Test_images)
sift_features_test_shape=np.shape(sift_features_test)

Test features: 380 * (49787, 128)


# 4. Extract features of training and test images using SURF (20 points)
Please print the size of extracted features, e.g., training features: 1149 * d, test features: 380 *d

In [90]:
import cv2

# Function to extract SURF features from training images
def extract_surf_train(train_images):
    # Create SURF object
    surf = cv2.xfeatures2d.SURF_create(extended=True)
    # Create an empty list to hold the descriptors of each image
    descriptors_list = []

    # Loop over each image in the training set
    for image in train_images:
        # Detect keypoints and compute SURF descriptors for the image
        kp, des = surf.detectAndCompute(image, None)
        # If the descriptors are not empty, add them to the list
        if des is not None:
            descriptors_list.append(des)

    # Concatenate the descriptors of all images in the list into a single numpy array
    descriptors_train = np.concatenate(descriptors_list, axis=0)

    # Print the size of the training features and the total number of features
    print("Training features:",len(train_images),'*', np.shape(descriptors_train))

    # Return the training descriptors
    return descriptors_train

# Function to extract SURF features from test images
def extract_surf_test(test_images):
    # Create SURF object
    surf = cv2.xfeatures2d.SURF_create(extended=True)
    # Create an empty list to hold the descriptors of each image
    descriptors_list = []

    # Loop over each image in the test set
    for image in test_images:
        # Detect keypoints and compute SURF descriptors for the image
        kp, des = surf.detectAndCompute(image, None)
        # If the descriptors are not empty, add them to the list
        if des is not None:
            descriptors_list.append(des)

    # Concatenate the descriptors of all images in the list into a single numpy array
    descriptors_test = np.concatenate(descriptors_list, axis=0)


    # Print the size of the test features and the total number of features
    print("Test features:",len(test_images),'*',np.shape(descriptors_test))


    # Return the test descriptors
    return descriptors_test


In [120]:
#getting SURF features from Train images
surf_features_train=extract_surf_train(Train_images)


Training features: 1149 * (478876, 128)


In [92]:
surf_features_train=extract_surf_train(Train_images)
surf_features_train_shape=np.shape(surf_features_train)

Training features: 1149 * (478876, 128)


In [93]:
#getting  SURF features from Test images
surf_features_test=extract_surf_test(Test_images)

Test features: 380 * (159158, 128)


In [94]:
surf_features_test=extract_surf_test(Test_images)
surf_features_test_shape=np.shape(surf_features_test)

Test features: 380 * (159158, 128)


# 5. Call SVM and kNN from scikit-learn and train the extracted HOG, SIFT and SURF features, respectively, save three CSV files of test dataset using three features (10 points)

In [ ]:
from sklearn.preprocessing import StandardScaler

# Create a StandardScaler object
scaler = StandardScaler()

# Fit the scaler to the training data and transform the data
#X_train_scaled = scaler.fit_transform(X_train)

# Use the scaler to transform the test data (don't fit it again!)
#X_test_scaled = scaler.transform(X_test)


In [96]:
# Reshape HOG features to have consistent number of samples
train_hog_features = np.array(hog_features_train)


# Convert the feature lists to numpy arrays
train_sift_features = np.array(sift_features_train, dtype=object)
train_surf_features = np.array(surf_features_train, dtype=object)
train_labels = np.array(train_labels).astype(int)

In [ ]:
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
import pandas as pd

# Train SVM and kNN classifiers on HOG features
svm_hog = SVC(kernel='linear')
svm_hog.fit(train_hog_features.reshape((-1, train_hog_features.shape[-1])), train_labels)

knn_hog = KNeighborsClassifier()
knn_hog.fit(train_hog_features.reshape((-1, train_hog_features.shape[-1])), train_labels)

# Train SVM and kNN classifiers on SIFT features
svm_sift = SVC(kernel='linear')
svm_sift.fit(np.vstack(train_sift_features), train_labels)

knn_sift = KNeighborsClassifier()
knn_sift.fit(np.vstack(train_sift_features), train_labels)

# Train SVM and kNN classifiers on SURF features
svm_surf = SVC(kernel='linear')
svm_surf.fit(np.vstack(train_surf_features), train_labels)

knn_surf = KNeighborsClassifier()
knn_surf.fit(np.vstack(train_surf_features), train_labels)


# Predict the labels of the test images using the SVM and kNN classifiers trained on HOG features
svm_hog_preds = svm_hog.predict(hog_test)
knn_hog_preds = knn_hog.predict(hog_test)

# Predict the labels of the test images using the SVM and kNN classifiers trained on SURF features
svm_surf_preds = svm_surf.predict(surf_test)
knn_surf_preds = knn_surf.predict(surf_test)

# Predict the labels of the test images using the SVM and kNN classifiers trained on SIFT features
svm_sift_preds = svm_sift.predict(sift_test)
knn_sift_preds = knn_sift.predict(sift_test)

# Save the test dataset and predicted labels to CSV files for each feature type
pd.DataFrame(hog_test).to_csv('hog_test.csv', index=False)
pd.DataFrame(svm_hog_preds).to_csv('svm_hog_preds.csv', index=False)
pd.DataFrame(knn_hog_preds).to_csv('knn_hog_preds.csv', index=False)

pd.DataFrame(surf_test).to_csv('surf_test.csv', index=False)
pd.DataFrame(svm_surf_preds).to_csv('svm_surf_preds.csv', index=False)
pd.DataFrame(knn_surf_preds).to_csv('knn_surf_preds.csv', index=False)

pd.DataFrame(sift_test).to_csv('sift_test.csv', index=False)
pd.DataFrame(svm_sift_preds).to_csv('svm_sift_preds.csv', index=False)
pd.DataFrame(knn_sift_preds).to_csv('knn_sift_preds.csv', index=False)


## Train SVM and kNN classifiers on the HOG features

In [100]:
# Train SVM and kNN classifiers on the HOG features
svm_hog = SVC(kernel='linear')
svm_hog.fit(hog_features_train, train_labels)
knn_hog = KNeighborsClassifier(n_neighbors=1)
knn_hog.fit(hog_features_train, train_labels)

KNeighborsClassifier(n_neighbors=1)

In [101]:
svm_hog_preds = svm_hog.predict(hog_features_test)
svm_hog_preds

array([0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1,
       0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0,
       1, 1, 1, 1, 0, 0, 0, 0, 2, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1,
       1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 1, 1, 2, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0,
       1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0,
       1, 2, 0, 1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 0,
       1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1,
       1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0,
       1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1,
       1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 2,
       1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1,

In [119]:
knn_hog_preds = knn_hog.predict(hog_features_test)
knn_hog_preds

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,

In [110]:
svm_hog_preds = svm_hog.predict(hog_features_test)
knn_hog_preds = knn_hog.predict(hog_features_test)

In [111]:
# Save the test dataset and predicted labels to CSV files for Hog feature
pd.DataFrame(hog_features_test).to_csv('hog_test.csv', index=False)
pd.DataFrame(svm_hog_preds).to_csv('svm_hog_preds.csv', index=False)
pd.DataFrame(knn_hog_preds).to_csv('knn_hog_preds.csv', index=False)


In [112]:
import pandas as pd

# Read hog_test.csv
hog_test_df = pd.read_csv('hog_test.csv')

# Read svm_hog_preds.csv
svm_hog_preds_df = pd.read_csv('svm_hog_preds.csv')

# Read knn_hog_preds.csv
knn_hog_preds_df = pd.read_csv('knn_hog_preds.csv')


In [114]:
svm_hog_preds_df.head()

0
0  0
1  1
2  0
3  0
4  1

In [118]:
len(knn_hog_preds_df)

380

In [116]:
hog_test_df.head()

0         1         2         3         4         5         6  \
0  0.240647  0.151494  0.080914  0.076415  0.174240  0.026870  0.042485   
1  0.050230  0.000000  0.025214  0.244225  0.345432  0.227832  0.000000   
2  0.234293  0.108554  0.219146  0.129896  0.232560  0.083012  0.000000   
3  0.294099  0.000000  0.021301  0.016840  0.000000  0.000000  0.010651   
4  0.237514  0.041598  0.195882  0.237514  0.141434  0.139827  0.237514   

          7         8         9  ...    142874    142875    142876    142877  \
0  0.146028  0.099182  0.240647  ...  0.053096  0.258879  0.166497  0.136957   
1  0.000000  0.000000  0.020092  ...  0.059436  0.264939  0.106691  0.264939   
2  0.099019  0.051450  0.234293  ...  0.093707  0.233696  0.030429  0.071871   
3  0.071148  0.000000  0.294099  ...  0.159333  0.236350  0.057950  0.073301   
4  0.041598  0.060612  0.193726  ...  0.025077  0.253851  0.062664  0.025804   

     142878    142879    142880    142881    142882    142883  
0  0.258879  0.258879  0.037545  0.101246  0.147245  0.051066  
1  0.095638  0.236590  0.069840  0.264939  0.166308  0.093073  
2  0.049443  0.233696  0.205840  0.143424  0.087044  0.017339  
3  0.046634  0.236350  0.073952  0.189966  0.132074  0.142732  
4  0.249275  0.115559  0.131172  0.108881  0.253851  0.156489  

[5 rows x 142884 columns]

# 6. Report the accuracy using Cow_teat_classfication_accuracy software, please attach the results image here (20 points)

### (1). SVM and 5NN using HOG features

### (2). SVM and 5NN using SIFT features

### (3). SVM and 5NN using SURF features